### PR TITLE
Feature/complete pull backends

### PR DESCRIPTION
**Pull request by gunther-bachmann**

### CODE REVIEW

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend [hunk 1]
```diff
@@ -60,53 +117,29 @@ class DocumentProvider {
...
-				doCleanCopy(file, newFile, pushAction)
-				copySuccessful = true;
-			} catch (Exception e) {
-				logger.error('exception during copy action', e)
-				copySuccessful = false;
-			}
-			if (!copySuccessful) {
-				logger.warn('resetting local repo to ' + preCommit.getName)
-				git.reset.setRef(preCommit.getName).setMode(ResetType.HARD).call
+			return [|
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:120]: shouldn't you be able to drop the pipe "|" if you don't list any parameters to the lambda…?

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend [hunk 2]
```diff
@@ -60,53 +117,29 @@ class DocumentProvider {
...
-		val pullResult = pull
-		val newCommitId = git.repository.resolve('HEAD')
-		if (newCommitId.equals(commitId) && pullResult.mergeResult.mergeStatus.successful) {
-			pushAction.apply(null)
-		} else {
-			val exceptionMessage = if (!newCommitId.equals(commitId)) {
-					'''unexpected inequality: pulled commit id '«newCommitId.getName»' != local commit id '«commitId.getName»' '''
-				} else if (!pullResult.mergeResult.mergeStatus.successful) {
-					'''merge conflicts in files «pullResult.mergeResult.conflicts.entrySet.map[key].toSet.join(', ')».'''
 				} else {
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:133]: I would suggest to just drop the "else" part

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend [hunk 3]
```diff
@@ -196,55 +203,42 @@ class DocumentProvider {
...
-		}
-		if (!result.equals(DocumentResource.createResult.succeeded)) {
-			logger.warn('resetting local repo to ' + preCommit.getName)
-			git.reset.setRef(preCommit.getName).setMode(ResetType.HARD).call
+			].wrapInCleanRepoActionWithCompensationOnFailure(pushAction, [
+				if (file.exists) {
+					// files not in the index, must be deleted, too (since resetting the index won't help)
+					logger.trace('''cleanup of remaining file '«file.absolutePath»' after reset hard necessary.''')
+					FileUtils.deleteQuietly(file)
+				} else {
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:232]: remove empty else block…?

**Review by jjelschen — APPROVED** [review]

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend (hunk 1 above)
Comment (gunther-bachmann) [reply org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:120]: in some cases that is true, but I encountered a couple of occurences where it had to be explicit so I figured I'll do it explicitly in all cases

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend (hunk 2 above)
Comment (gunther-bachmann) [reply org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:133]: I had an eclipse warning if else block is missing

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend (hunk 3 above)
Comment (gunther-bachmann) [reply org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:232]: I had an eclipse warning if else block is missing

**Pull request merged**